dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -663,6 +663,322 @@ template class NaryOpNode<functional::min<double>>;
 template class NaryOpNode<std::multiplies<double>>;
 template class NaryOpNode<std::plus<double>>;
 
+// PartialReduceNode *****************************************************************
+std::vector<ssize_t> partial_reduce_shape(const std::span<const ssize_t> input_shape,
+                                          const ssize_t axis) {
+    std::vector<ssize_t> shape;
+    shape.assign(input_shape.begin(), input_shape.end());
+    shape.erase(shape.begin() + axis);
+    return shape;
+}
+
+std::vector<ssize_t> as_contiguous_strides(const std::span<const ssize_t> shape) {
+    ssize_t ndim = static_cast<ssize_t>(shape.size());
+
+    assert(ndim >= 0);
+    std::vector<ssize_t> strides(ndim);
+    // otherwise strides are a function of the shape
+    strides[ndim - 1] = sizeof(double);
+    for (auto i = ndim - 2; i >= 0; --i) {
+        strides[i] = strides[i + 1] * shape[i + 1];
+    }
+    return strides;
+}
+
+/// TODO: support multiple axes
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* node_ptr, std::span<const ssize_t> axes,
+                                               double init)
+        : ArrayOutputMixin(partial_reduce_shape(node_ptr->shape(), axes[0])),
+          init(init),
+          array_ptr_(node_ptr),
+          axes_(make_axes(axes)),
+          parent_strides_c_(as_contiguous_strides(array_ptr_->shape())) {
+    if (array_ptr_->dynamic()) {
+        throw std::invalid_argument("cannot do a partial reduction on a dynamic array");
+    } else if (array_ptr_->size() < 1) {
+        throw std::invalid_argument("cannot do a partial reduction on an empty array");
+    }
+
+    // Validate axes
+    /// TODO: support negative axes
+    assert(this->ndim() == array_ptr_->ndim() - 1);
+    ssize_t axis = axes_[0];
+    if (axis < 0 || axis >= array_ptr_->ndim()) {
+        throw std::invalid_argument("Axes should be integers between 0 and n_dim - 1");
+    }
+
+    add_predecessor(node_ptr);
+}
+
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* node_ptr,
+                                               std::initializer_list<ssize_t> axes, double init)
+        : PartialReduceNode(node_ptr, std::span(axes), init) {}
+
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* node_ptr, const ssize_t axis, double init)
+        : PartialReduceNode(node_ptr, {axis}, init) {}
+
+template <>
+PartialReduceNode<std::plus<double>>::PartialReduceNode(ArrayNode* array_ptr, const ssize_t axis)
+        : PartialReduceNode(array_ptr, axis, 0) {}
+
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* array_ptr, std::span<const ssize_t> axes)
```

Comment:
I don't think we want this generic overload, right? It doesn't set up `init` and we already cover the one case that does.

Reply:
so remove that also from the header?

Reply:
So the header file keeps it. But we only implement it for the relevant type overloads.